### PR TITLE
OAK-10254 : added a test to reproduce late-write inconsistencies

### PR DESCRIPTION
**Pull request by stefan-egli**

### CODE REVIEW

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/UnrecoveredRevisionTest.java [hunk 1]
```diff
@@ -0,0 +1,813 @@
...
+            NodeState n = ns.getRoot();
+            for (String name : p.elements()) {
+                n = n.getChildNode(name);
+            }
+            System.out.println(aPath + " :");
+            if (!n.exists()) {
+                System.out.println("DOES NOT EXIST");
+            } else {
+                DocumentNodeState dns = (DocumentNodeState) n;
+                NodeDocument doc = dns.getDocument();
```
Comment (mreutegg) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/UnrecoveredRevisionTest.java]: This method does not exist. Rather get the path from the DocumentNodeState and then look up the document from the DocumentStore?

**Review by mreutegg — CHANGES_REQUESTED** [review]

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/UnrecoveredRevisionTest.java [hunk 2]
```diff
@@ -0,0 +1,813 @@
...
+        public <T extends Document> void remove(Collection<T> collection, String key) {
+            allowOp(collection);
+            super.remove(collection, key);
+        }
+        
+        @Override
+        public <T extends Document> int remove(Collection<T> collection,
+                String indexedProperty, long startValue, long endValue)
+                throws DocumentStoreException {
+            throw new IllegalStateException("not yet implemented");
```
Comment (rishabhdaim) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/UnrecoveredRevisionTest.java]: Shouldn't this be `UnsupportedOperationException`,  wdyt?

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/UnrecoveredRevisionTest.java [hunk 3]
```diff
@@ -0,0 +1,813 @@
...
+    private Clock virtualClock;
+    private DocumentNodeStore ns1;
+    private DocumentNodeStore ns2;
+
+    private ControllableMemoryDocumentStore store;
+    
+    @Before
+    public void setup() throws Exception {
+        int createOrUpdateBatchSize = 1;
+        setup(createOrUpdateBatchSize);
```
Comment (rishabhdaim) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/UnrecoveredRevisionTest.java]: we can inline this parameter and remove local variable.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/UnrecoveredRevisionTest.java [hunk 4]
```diff
@@ -0,0 +1,813 @@
...
+                writtenUpdateOps.addAll(trimmedOps);
+            }
+        }
+
+        private void setBreakpointCallback(BreakpointCallback callback) {
+            this.callback = callback;
+        }
+        
+        @SuppressWarnings("unused")
+        private void setRestrictedCollections(Collection<?>... collections) {
```
Comment (rishabhdaim) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/UnrecoveredRevisionTest.java:111]: Can we remove this un-used method or do we plan to use it in the future?

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/UnrecoveredRevisionTest.java (hunk 1 above)
Comment (stefan-egli) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/UnrecoveredRevisionTest.java]: oups, missed to clean that up, had that locally hacked in.

now fixed in https://github.com/apache/jackrabbit-oak/pull/950/commits/3846d96be9eb30904a1bc836c8e968525f579faf

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/UnrecoveredRevisionTest.java (hunk 2 above)
Comment (stefan-egli) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/UnrecoveredRevisionTest.java]: Yes looks more appropriate indeed. I've now changed those in commit : https://github.com/apache/jackrabbit-oak/pull/950/commits/ced8a85c86475d270d2bc2d230c2c344a6e52ac8

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/UnrecoveredRevisionTest.java (hunk 3 above)
Comment (stefan-egli) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/UnrecoveredRevisionTest.java]: yes, looks a bit ugly indded, I've fixed this now in https://github.com/apache/jackrabbit-oak/pull/950/commits/39132a3385dd1d30d0e89d091d5f75e0b27f4523

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/UnrecoveredRevisionTest.java (hunk 4 above)
Comment (stefan-egli) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/UnrecoveredRevisionTest.java:111]: I'd prefer to have it in for now? I was expecting (or let's say hoping) it to eventually be used, once we go more fancy with test coverage.

**Review by rishabhdaim — APPROVED** [review]

**Pull request merged**